Sources/NIOTransportServices/NIOTSConnectionBootstrap.swift

```diff
@@ -261,6 +274,528 @@ public final class NIOTSConnectionBootstrap {
+                    switch result {
+                    case .success:
+                        connectionChannel.triggerUserOutboundEvent(
+                            NIOTSNetworkEvents.ConnectToNWEndpoint(endpoint: endpoint),
+                            promise: promise
+                        )
+                    case .failure(let error):
+                        promise.fail(error)
+                    }
+                }
+            },
+            postRegisterTransformation: { $1.makeSucceededFuture($0) }
+        ).get()
+    }
+
+    /// Use a pre-existing `NWConnection` to connect a `Channel`.
+    ///
+    /// - Parameters:
+    ///   - connection: The `NWConnection` to wrap.
+    ///   - channelInitializer: A closure to initialize the channel. The return value of this closure is returned from the `connect`
+    ///   method.
+    /// - Returns: The result of the channel initializer.
+    @available(macOS 10.15, iOS 13, tvOS 13, watchOS 6, *)
+    @_spi(AsyncChannel)
+    public func withExistingNWConnection<Output: Sendable>(
+        _ connection: NWConnection,
+        channelInitializer: @escaping @Sendable (Channel) -> EventLoopFuture<Output>
+    ) async throws -> Output {
+        try await self.connect0(
+            existingNWConnection: connection,
+            channelInitializer: channelInitializer,
+            registration: { connectionChannel, promise in
+                connectionChannel.registerAlreadyConfigured0(promise: promise)
+            },
+            postRegisterTransformation: { $1.makeSucceededFuture($0) }
+        ).get()
+    }
+
+    @available(macOS 10.15, iOS 13, tvOS 13, watchOS 6, *)
+    private func connect0<ChannelInitializerResult, PostRegistrationTransformationResult>(
+        existingNWConnection: NWConnection? = nil,
+        channelInitializer: @escaping @Sendable (Channel) -> EventLoopFuture<ChannelInitializerResult>,
+        registration: @escaping (NIOTSConnectionChannel, EventLoopPromise<Void>) -> Void,
+        postRegisterTransformation: @escaping @Sendable (ChannelInitializerResult, EventLoop) -> EventLoopFuture<PostRegistrationTransformationResult>
+    ) -> EventLoopFuture<PostRegistrationTransformationResult> {
+        let connectionChannel: NIOTSConnectionChannel
+        if let newConnection = existingNWConnection {
+            connectionChannel = NIOTSConnectionChannel(
+                wrapping: newConnection,
+                on: self.group.next() as! NIOTSEventLoop,
+                tcpOptions: self.tcpOptions,
+                tlsOptions: self.tlsOptions
+            )
+        } else {
+            connectionChannel = NIOTSConnectionChannel(
+                eventLoop: self.group.next() as! NIOTSEventLoop,
+                qos: self.qos,
+                tcpOptions: self.tcpOptions,
+                tlsOptions: self.tlsOptions
+            )
+        }
+        let channelInitializer = { (channel: Channel) -> EventLoopFuture<ChannelInitializerResult> in
```

Comment:
It's amassing how many layers of wrapping are here when you remember the protocol handlers 😄

Reply:
Don't get me started 🙃